File: video-uploader/src/main/java/org/breizhcamp/video/uploader/services/EventSrv.java
package org.breizhcamp.video.uploader.services;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.breizhcamp.video.uploader.dto.Event;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.util.List;
import java.util.Map;

import static java.util.function.Function.identity;
import static java.util.stream.Collectors.toMap;

@Service
public class EventSrv {

	@Autowired
	private ObjectMapper mapper;

	/**
	 * List all events in schedule
	 * @return Event list
	 */
	public List<Event> list() throws IOException {
		InputStream schedule = EventSrv.class.getResourceAsStream("/schedule.json");
		return mapper.readValue(schedule, new TypeReference<List<Event>>(){});
	}

	/**
	 * Retrieve all events group by id
	 * @return Event list
	 */
	public Map<Integer, Event> listById() throws IOException {
		return list().stream()
				.collect(toMap(Event::getId, identity()));
	}

	/**
	 * Retrieve an event with his id
	 * @param id Id of the event to retrieve
	 * @return Event
	 * @throws IOException If we cannot read schedule json file
	 * @throws FileNotFoundException If the event is not found in schedule
	 */
	public Event getFromId(Integer id) throws IOException {
		return list().stream()
				.filter(e -> e.getId().equals(id))
				.findFirst()
				.orElseThrow(() -> new FileNotFoundException("Event [" + id + "] not found in schedule"));
	}
}

File: src/main/java/org/breizhcamp/video/uploader/services/EventSrv.kt
package org.breizhcamp.video.uploader.services

import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.fasterxml.jackson.module.kotlin.jacksonMapperBuilder
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import org.breizhcamp.video.uploader.CamaalothUploaderProps
import org.breizhcamp.video.uploader.dto.Event
import org.springframework.stereotype.Service
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.lang.Integer.max
import java.nio.file.Paths

@Service
class EventSrv(props: CamaalothUploaderProps) {

    val mapper = jacksonMapperBuilder().addModule(JavaTimeModule()).build()

    val scheduleFile: File = Paths.get(props.assetsDir, "schedule.json").toFile()

    /**
     * Read all events in schedule file
     * @return Event list
     */
    @Throws(IOException::class)
    fun read() = mapper.readValue<List<Event>>(scheduleFile)

    /**
     * Retrieve all events group by id
     * @return Event list
     */
    @Throws(IOException::class)
    fun readAndAssociateByIds() = read().associateBy { it.id }

    /**
     * Retrieve an event with his id
     * @param id Id of the event to retrieve
     * @return Event
     * @throws IOException If we cannot read schedule json file
     * @throws FileNotFoundException If the event is not found in schedule
     */
    @Throws(IOException::class)
    fun readAndGetById(id: Int): Event? = readAndAssociateByIds()[id]

    /**
     * Generate ids for the events that don't have one
     */
    fun generateMissingIdsAndWrite() {
        val events = read()
        var nextId = events.fold(0) { acc, event -> max(acc, event.id ?: 0) } + 10000000

        val fixedEvents = events.map { event ->
            if (event.id == null) {
                event.id = nextId
                nextId++
            }

            event
        }

        mapper.writeValue(scheduleFile, fixedEvents)
    }
}